quickwit-indexing/src/source/kinesis/kinesis_source.rs

```diff
@@ -0,0 +1,530 @@
+// Copyright (C) 2021 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::fmt;
+
+use anyhow::Context;
+use async_trait::async_trait;
+use itertools::Itertools;
+use quickwit_actors::{ActorExitStatus, Mailbox};
+use quickwit_config::KinesisSourceParams;
+use quickwit_metastore::checkpoint::{CheckpointDelta, PartitionId, Position, SourceCheckpoint};
+use rusoto_core::Region;
+use rusoto_kinesis::KinesisClient;
+use serde_json::json;
+use tokio::sync::mpsc;
+use tokio::time;
+use tracing::{info, warn};
+
+use super::api::list_shards;
+use super::shard_consumer::{ShardConsumer, ShardConsumerHandle, ShardConsumerMessage};
+use crate::models::RawDocBatch;
+use crate::source::{IndexerMessage, Source, SourceContext, TypedSourceFactory};
+
+const TARGET_BATCH_NUM_BYTES: u64 = 5_000_000;
+
+type ShardId = String;
+
+/// Factory for instantiating a `KafkaSource`.
+pub struct KinesisSourceFactory;
+
+#[async_trait]
+impl TypedSourceFactory for KinesisSourceFactory {
+    type Source = KinesisSource;
+    type Params = KinesisSourceParams;
+
+    async fn typed_create_source(
+        params: KinesisSourceParams,
+        checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self::Source> {
+        KinesisSource::try_new(params, checkpoint).await
+    }
+}
+
+struct ShardConsumerState {
+    partition_id: PartitionId,
+    position: Position,
+    lag_millis: Option<i64>,
+    shard_consumer_handle: ShardConsumerHandle,
+}
+
+#[derive(Default)]
+pub struct KinesisSourceState {
+    /// Pool of [`ShardConsumer`] managed by the source.
+    shard_consumers: HashMap<ShardId, ShardConsumerState>,
+    /// Number of bytes processed by the source.
+    pub num_bytes_processed: u64,
+    /// Number of records processed by the source (including invalid messages).
+    pub num_records_processed: u64,
+    // Number of invalid records, i.e., that were empty or could not be parsed.
+    pub num_invalid_records: u64,
+}
+
+pub struct KinesisSource {
+    stream_name: String,
+    checkpoint: SourceCheckpoint,
+    kinesis_client: KinesisClient,
+    sink_tx: mpsc::Sender<ShardConsumerMessage>,
```

Comment:
This channel is used to communicate with the shards? A rename or a comment would be helpful.